### PR TITLE
177911305 improve stake concentration score

### PR DESCRIPTION
**Pull request by pawjer**
#### What's this PR do?
- Add queries to handle class query methods
- Calculate top 33% of stake possesion

#### How should this be manually tested?
- Run tests

#### What are the relevant tickets?
https://www.pivotaltracker.com/story/show/177911305

#### Task completed checklist:
- [x] Is there appropriate test coverage?
- [x] Did I take into consideration possible security vulnerabilities?
- [x] Are the controllers secure?
- [x] Was every important matter documented?

#### Things to watch out for
- [N] Does this PR have any migrations?
- [N] Does this PR add new dependencies?

### CODE REVIEW

**Comment by kbiala** [comment]
@pawjer Please also update Stake Concentration Score section in FAQ: https://www.validators.app/faq.

File: app/controllers/api/v1/api_controller.rb [hunk 1]
```diff
@@ -61,11 +61,12 @@ def ping_times
 
       # Show the list of validators with scores
       def validators_list
-        @sort_order = if params[:order] == 'score'
+        @sort_order = case params[:order]
```
Comment (maciejkocylapc) [inline app/controllers/api/v1/api_controller.rb:64]: can you use index_order method from validator model instead ?

**Review by maciejkocylapc — APPROVED** [review]

File: app/controllers/public_controller.rb [hunk 2]
```diff
@@ -138,6 +109,13 @@ def contact_us
 
   private
 
+  def validate_order
+    valid_orders = %w[score name stake random]
```
Comment (radek-polcode) [inline app/controllers/public_controller.rb:113]: Maybe create a constant in the model?

File: app/controllers/validators_controller.rb [hunk 3]
```diff
@@ -7,64 +7,47 @@ class ValidatorsController < ApplicationController
...
-    @skipped_vote_percent_best = \
-      (@slot_index_current - @credits_current_max )/@slot_index_current.to_f
+    @skipped_vote_percent_best = vote_account_history_query.skipped_vote_percent_best
 
-    # flash[:error] = 'Due to a problem with our RPC server pool, the Skipped Slot % data is inaccurate. I am aware of the problem and working on a better solution. Thanks, Brian Long'
+    validator_history =
+      ValidatorHistoryQuery.new(params[:network], @batch.uuid)
+
+    at_33_stake_validator = validator_history.at_33_stake&.validator
+    @at_33_stake_index = (validators.index(at_33_stake_validator)&.+ 1).to_i
```
Comment (radek-polcode) [inline app/controllers/validators_controller.rb:48]: What is the reason you use safe operator? There is possibility of nil? If so is it ok?
```
2.6.5 :008 > nil&.+1.to_i
 => nil 
```

File: app/controllers/validators_controller.rb [hunk 4]
```diff
@@ -7,64 +7,47 @@ class ValidatorsController < ApplicationController
...
 
+    # flash[:error] = 'Due to a problem with our RPC server pool, the Skipped Slot % data is inaccurate. I am aware of the problem and working on a better solution. Thanks, Brian Long'
```
Comment (radek-polcode) [inline app/controllers/validators_controller.rb:50]: Do we want to keep this comment?

File: app/helpers/validators_helper.rb [hunk 5]
```diff
@@ -43,7 +42,12 @@ def current_software_version(batch, network)
   def skipped_vote_percent(validator, batch, skipped_vote_percent_best)
     vahl = validator.vote_account_last&.vote_account_history_for(batch.uuid)
     return unless vahl
+
     skipped_votes_percent = (vahl.slot_index_current.to_i - vahl.credits_current.to_i)/vahl.slot_index_current.to_f
     ((skipped_vote_percent_best - skipped_votes_percent.to_f)*100.0).round(2)
   end
+
+  def above_33percent_concentration?(validator)
```
Comment (radek-polcode) [inline app/helpers/validators_helper.rb:50]: Wouldn't it be better to have it in the model?

File: app/helpers/validators_helper.rb [hunk 6]
```diff
@@ -43,7 +42,12 @@ def current_software_version(batch, network)
   def skipped_vote_percent(validator, batch, skipped_vote_percent_best)
```
Comment (radek-polcode) [inline app/helpers/validators_helper.rb:42]: Wouldn't it be better to have it in the model?

File: test/queries/validator_history_query_test.rb [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+    @root_blocks = @validator_histories[0...-1].map(&:root_block)
+    @last_votes = @validator_histories[0...-1].map(&:last_vote)
+    @active_stakes = @validator_histories[0...-1].map(&:active_stake)
+  end
+
+  def teardown
+    super
+
+    @validator_histories.each(&:destroy)
```
Comment (radek-polcode) [inline test/queries/validator_history_query_test.rb:39]: `delete_all` maybe?

File: app/models/batch.rb [hunk 8]
```diff
@@ -31,6 +31,10 @@
 class Batch < ApplicationRecord
   before_create :create_uuid
 
+  def self.last_scored(network)
+    where('network = ? and scored_at IS NOT NULL', network).last
```
Comment (radek-polcode) [inline app/models/batch.rb:35]: Don't you need some ordering here or last is enough?

File: app/controllers/public_controller.rb (hunk 2 above)
Comment (pawjer) [reply app/controllers/public_controller.rb:113]: Nope - it's different for API, public, and validator controllers.

File: app/controllers/validators_controller.rb (hunk 3 above)
Comment (pawjer) [reply app/controllers/validators_controller.rb:48]: You've used it wrong. You should try as follows:
```
2.6.5 :001 > (nil&._1).to_i
 => 0
 ```

File: app/controllers/validators_controller.rb (hunk 4 above)
Comment (pawjer) [reply app/controllers/validators_controller.rb:50]: It's not about my changes. So I didn't remove it.

File: app/helpers/validators_helper.rb (hunk 5 above)
Comment (pawjer) [reply app/helpers/validators_helper.rb:50]: It's related to the view, so I don't think so.

File: app/models/batch.rb (hunk 8 above)
Comment (pawjer) [reply app/models/batch.rb:35]: Default is the order by id, so if I understand it correctly it should always point to the newly created object.

File: app/models/validator.rb [hunk 9]
```diff
@@ -73,12 +97,12 @@ def ip_address
   end
 
   def copy_data_to_score
-    if validator_score_v1
-      validator_score_v1.ip_address = ip_address
-      ip_dc = Ip.where(address: ip_address).first&.data_center_key
-      validator_score_v1.data_center_key = ip_dc
-      validator_score_v1.save
-    end
+    return unless validator_score_v1
```
[the remark this answers is not in the document]
Comment (pawjer) [reply app/models/validator.rb:100]: Could you explain to me your thought a bit deeper?

File: app/helpers/validators_helper.rb (hunk 6 above)
Comment (pawjer) [reply app/helpers/validators_helper.rb:42]: It's not about my changes.

File: test/queries/validator_history_query_test.rb (hunk 7 above)
Comment (pawjer) [reply test/queries/validator_history_query_test.rb:39]: There's no such a method for an Array.

**Pull request merged**